### PR TITLE
[internal/otelarrow] Resolve test flakes; skip one still-flaky test

### PR DESCRIPTION
**Pull request by jmacd**
**Description:** Fixes the causes of flakiness in most cases by using a callback to terminate the test without resorting to sleep statements. There is still one flaky test that for reasons not understood, does not pass. Fortunately, it fails in a repeatable way, and I will debug as part of #34719. 

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/34719

### CODE REVIEW

**Review by mwear — COMMENTED** [review]
Can we leave the test active in non-windows environments. Something like:

```go
if runtime.GOOS == "windows" {
		t.Skip("skipping flaky test on Windows, see https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/34719")
	}
```

**Comment by jmacd** [comment]
This test is failing for other reasons in my local environment. The local failure was an expectation that more than one stream was used, and I find the test wasn't controlling for that variable. I introduced retries for thoroughness and a tiny sleep call to ensure the expected >1 streams takes place. It can be no worse for the Windows problem, and I see it has passed at least once on Windows (from the history above, though see other test flakes on Windows).

**Review by mwear — APPROVED** [review]
The failing test is an instance of: #34792

**Review by pjanotti — APPROVED** [review]
I'm ignoramus about the internals of this component, but, when testing in Windows, adding a sleep can be advisable given the differences in scheduling and time tick resolution in relation *nix where most devs run their tests. LGTM.

File: internal/otelarrow/test/e2e_test.go [hunk 1]
```diff
@@ -194,12 +202,16 @@ func testIntegrationTraces(ctx context.Context, t *testing.T, tp testParams, cfg
...
 
+	slp := time.Duration(float64(testCon.expCfg.Arrow.MaxStreamLifetime) * 3 / 2 / float64(tp.requestCount))
 	for num := 0; num < tp.threadCount; num++ {
 		clientDoneWG.Add(1)
 		go func(num int) {
 			defer clientDoneWG.Done()
 			generator := mkgen()
 			for i := 0; i < tp.requestCount; i++ {
+				// Make this process take 1.5x the stream lifetime
+				time.Sleep(slp)
```
Comment (jpkrohling) [inline internal/otelarrow/test/e2e_test.go:213]: is it possible to wait on a given condition, instead of waiting on time?

**Review by jpkrohling — CHANGES_REQUESTED** [review]

File: internal/otelarrow/test/e2e_test.go (hunk 1 above)
Comment (jmacd) [reply internal/otelarrow/test/e2e_test.go]: Yes. Fixed!
The tests in this file use a callback to signal the end of the test.

**Comment by jmacd** [comment]
There is something suspicious about the failures which makes me think the test itself is not responsible. As you see, I've enabled a line of code that causes logs to print to the console, which shows the error the test is expecting. Somehow the zap Observer is failing, and I have a feeling it relates to Windows.

**Comment by crobert-1** [comment]
_I've submitted https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/34823 for the `receiver/datadog` test failures, unrelated to this change._

**Comment by jmacd** [comment]
There is still a flaky test, the problem appears to be a real problem combined with a slower test start than on the non-windows host. I have added a `Skip()` and will leave https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/34719 open to cover fixing it.

**Comment by jmacd** [comment]
I recommend we merge this as-is, because (a) it will take some time to properly fix the one skipped test, (b) I have four PRs open in this repository and it becomes difficult to move anything in this situation.

**Review by jpkrohling — APPROVED** [review]
Nice, thanks!

**Pull request merged**